File: python/src/multiagent_boilerplate/harness/schemas.py
"""Loads specs/ files - the single source of truth shared with the TypeScript runtime."""

from __future__ import annotations

import json
from functools import cache
from pathlib import Path


def _find_repo_root(start_dir: Path) -> Path:
    """Walks up from this module's actual location until it finds `specs/schemas` -
    deliberately not a fixed parent-count, so this keeps working regardless of how deep this
    file sits relative to the repo root. A deployment must keep `specs/` as a sibling of
    `python/` on disk either way - see deploy/Dockerfile.python - this just stops assuming a
    specific directory depth to find it."""
    directory = start_dir
    while True:
        if (directory / "specs" / "schemas").is_dir():
            return directory
        if directory.parent == directory:
            raise RuntimeError(f"could not locate repo root (no ancestor of {start_dir} contains specs/schemas)")
        directory = directory.parent


REPO_ROOT = _find_repo_root(Path(__file__).resolve().parent)
SCHEMAS_DIR = REPO_ROOT / "specs" / "schemas"
AGENTS_DIR = REPO_ROOT / "specs" / "agents"


@cache
def load_schema(file_name: str) -> dict:
    return json.loads((SCHEMAS_DIR / file_name).read_text(encoding="utf-8"))


@cache
def load_agent_role(role_name: str) -> dict:
    return json.loads((AGENTS_DIR / f"{role_name}.json").read_text(encoding="utf-8"))


@cache
def load_prompt(prompt_relative_path: str) -> str:
    return (REPO_ROOT / prompt_relative_path).read_text(encoding="utf-8")

File: typescript/src/harness/schemas.ts
import { existsSync, readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

/**
 * Walks up from this module's actual location until it finds `specs/schemas` - deliberately
 * not a fixed parent-count, since that breaks the moment this file is bundled (tsup flattens
 * src/harness/schemas.ts into dist/, changing its depth relative to the repo root). A
 * deployment must keep `specs/` as a sibling of `typescript/` on disk either way - see
 * deploy/Dockerfile.typescript - this just stops assuming a specific directory depth to find it.
 */
function findRepoRoot(startDir: string): string {
  let dir = startDir;
  for (;;) {
    if (existsSync(resolve(dir, "specs", "schemas"))) return dir;
    const parent = dirname(dir);
    if (parent === dir) {
      throw new Error(`could not locate repo root (no ancestor of ${startDir} contains specs/schemas)`);
    }
    dir = parent;
  }
}

const repoRoot = findRepoRoot(dirname(fileURLToPath(import.meta.url)));
const schemasDir = resolve(repoRoot, "specs", "schemas");

export function loadSchema(fileName: string): Record<string, unknown> {
  const path = resolve(schemasDir, fileName);
  return JSON.parse(readFileSync(path, "utf-8")) as Record<string, unknown>;
}

export function loadAgentRole(roleName: string): Record<string, unknown> {
  const path = resolve(repoRoot, "specs", "agents", `${roleName}.json`);
  return JSON.parse(readFileSync(path, "utf-8")) as Record<string, unknown>;
}

export function loadPrompt(promptRelativePath: string): string {
  const path = resolve(repoRoot, promptRelativePath);
  return readFileSync(path, "utf-8");
}

export const repoRootPath = repoRoot;
